flink-kubernetes/src/main/java/org/apache/flink/kubernetes/KubernetesClusterClientFactory.java

```diff
@@ -51,10 +51,7 @@ public boolean isCompatibleWith(Configuration configuration) {
 	@Override
 	public KubernetesClusterDescriptor createClusterDescriptor(Configuration configuration) {
 		checkNotNull(configuration);
-		if (!configuration.contains(KubernetesConfigOptions.CLUSTER_ID)) {
-			final String clusterId = generateClusterId();
-			configuration.setString(KubernetesConfigOptions.CLUSTER_ID, clusterId);
-		}
+		ensureClusterIdIsSet(configuration);
 		return new KubernetesClusterDescriptor(configuration, KubeClientFactory.fromConfiguration(configuration));
```

Comment:
It is a bit unrelated. But setting a fixed cluster id for the `ClusterDescriptor` seems wrong. That way we can ever deploy a single cluster using a `ClusterDescriptor`. I believe that the `YarnClusterDescriptor` works differently.

Reply:
Yes I noticed that too, it is at odds with the interface. You can connect to _any_ cluster, but only deploy a single one.
But this seems to be an issue of the `ClusterDescriptor` interface; there is no way for implementations to accept additional arguments for deploying clusters.

Reply:
The YarnClusterDescriptor actually works quite similarly; there things like custom names and application types are being defined in the constructor.
The `ClusterDescriptor` just doesn't handle this case very well, but admittedly we do not really have a use-case for deploying multiples clusters.

Reply:
I think the idea was that the `ClusterDescriptor` will generate a cluster id when you call any of the `deploy*` methods. I believe what is set in the constructor of the `YarnClusterDescriptor` is fine and does not contradict what I was saying.

If one really needs to configure the cluster id manually, then one could extend the `ClusterDescriptor` interface accordingly.

I think that we are adding technical debt by working around broken contracts which might be fine now but we should definitely pull this straight eventually.